solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -699,7 +699,10 @@ private void constructState(Set<String> changedCollections) {
           clusterState.getCollectionStates());
     }
 
-    notifyCloudCollectionsListeners();
+    if (changedCollections.isEmpty()) {
+      // We only need to notify cloud collection listeners on new collections
+      notifyCloudCollectionsListeners();
```

Comment:
That particular callback meant for number of collection change only(collection create or delete). It doesn't care about collection update. (May be just update the comments there). I wish we have separate `consturctState` method for collection-create/delete vs collection-update. That would have clearly isolate the difference. 

In any case it should help in our case. We can refactor zkStateReader and optimize this clusterstate further.